run_gunicorn.sh

```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+echo "Launching gunicorn application..."
+poetry run flask run --host=0.0.0.0
+#poetry run gunicorn --bind 0.0.0.0:5000 wsgi:app
```

Comment:
We want to make sure our production container is running with Gunicorn (even the flask logs advise not to use it in production!) so worth us investigating why this isn't working. I think the two steps we need are:
* Track Gunicorn with poetry so that it's installed by default within our container
* Adjust the starting path:
  * Gunicorn wants the actual app variable returned by `create_app` whereas `app.py` only bothers to instantiate one if we call it directly (`if __name__ == '__main__':`) - so this won't quite play ball. Instead we can more easily grab such a variable with the function directly: `"app:create_app()"`
  * But we also need to consider the module structure that we're using, so we need full references, i.e. this would be actually be `todo_app.app:create_app()"`

Reply:
Please Help @JackMead 

I still can't get the gunicorn command syntax right...

gunicorn --bind 0.0.0.0:5000 todo_app.app:create_app

```
Traceback (most recent call last):
  File "c:\users\jacki\appdata\local\programs\python\python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\users\jacki\appdata\local\programs\python\python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\jacki\AppData\Local\Programs\Python\Python39\Scripts\gunicorn.exe\__main__.py", line 4, in <module>
  File "c:\users\jacki\appdata\local\programs\python\python39\lib\site-packages\gunicorn\app\wsgiapp.py", line 9, in <module>
    from gunicorn.app.base import Application
  File "c:\users\jacki\appdata\local\programs\python\python39\lib\site-packages\gunicorn\app\base.py", line 11, in <module>
    from gunicorn import util
  File "c:\users\jacki\appdata\local\programs\python\python39\lib\site-packages\gunicorn\util.py", line 8, in <module>
    import fcntl
ModuleNotFoundError: No module named 'fcntl'
```

Reply:
Good news and Bad news.

**poetry run gunicorn --bind 0.0.0.0:5000 todo_app.app:create_app**

When I run through Docker it now launches the Gunicorn application "successfully":

```
PS C:\Work\Module5\DevOps-Course-Starter> docker run --env-file ./.env -it --publish 5000:5000 do-me:prod
Launching application from Gunicorn...
[2021-10-31 21:06:16 +0000] [7] [INFO] Starting gunicorn 20.1.0
[2021-10-31 21:06:16 +0000] [7] [INFO] Listening at: http://0.0.0.0:5000 (7)
[2021-10-31 21:06:16 +0000] [7] [INFO] Using worker: sync
[2021-10-31 21:06:16 +0000] [11] [INFO] Booting worker with pid: 11
```

But when I try to browse it gives an error:

```
[2021-10-31 21:06:28 +0000] [11] [ERROR] Error handling request /
Traceback (most recent call last):
  File "/root/.cache/pypoetry/virtualenvs/todo-app-P8TiTxj2-py3.8/lib/python3.8/site-packages/gunicorn/workers/sync.py", line 136, in handle
    self.handle_request(listener, req, client, addr)
  File "/root/.cache/pypoetry/virtualenvs/todo-app-P8TiTxj2-py3.8/lib/python3.8/site-packages/gunicorn/workers/sync.py", line 179, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
TypeError: create_app() takes 0 positional arguments but 2 were given
```

Reply:
FIXED IT!

Didn't realise I have to escape the brackets in bash shell script:

`poetry run gunicorn --bind 0.0.0.0:5000 todo_app.app:create_app\(\)`